Source/NSubstitute/Core/ICustomHandlers.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+
+namespace NSubstitute.Core
+{
+    /// <summary>
+    ///     Factory method which creates <see cref="ICallHandler" /> from the <see cref="ISubstituteState" />.
+    /// </summary>
+    public delegate ICallHandler CallHandlerFactory(ISubstituteState substituteState);
+
+    public interface ICustomHandlers
+    {
+        IEnumerable<ICallHandler> Handlers { get; }
+        IDictionary<object, object> HandlerDataStorage { get; }
```

Comment:
Well, yes, AutoFixture will store its "known" results there - each subsequent call to same method should return the same result. That results are per-substitute, so it's an ideal place to store them.

Do you see any other place where AF can store cached results for substitute calls? I don't see, unfortunately :( Keeping a custom map (`subsitute object` to `AF state`) looks a horrible idea.

Another way is to guarantee that we always create a single instance of `CallHandler` per substitute. But is that OK to make such guarantee a part of `CallHandler` design? That would be the only `ICallHandler` we make such guarantee for..